Resolve hostnames

dgalaktionov opened the pull request:
Resolves hostnames (such as `localhost` or `github.com`) for endpoint addresses. Gives preference to IPv4 over IPv6, although it works with both.

This is a small first contribution where my main goal was to be able to build the project and familiarize with the code. Feel free to point out any issues you may see!

dgalaktionov committed:
Resolve hostnames

yamashi commented:
Hello, thank you for the PR.
I have a few observations:

- The code guidelines are not followed (but that's not your fault, I have yet to write them down), in this instance it's the argument names, they need to be prefixed with an 'a' + 'c' if const + 'p' if pointer for example: void Test(const int* acpSomeValue); 

- The endpoint class should not be the one responsible for name resolution, we probably need a dedicated class for this "Resolver" that could be started asynchronously and contain a list of all the endpoints after resolution. On the top of my head I am thinking it could be a great use of std::async and std::future.

- The for loop used to set the hostname will loop until the last element is reached, could break the loop once you find the ipv4 entry you are looking for.

Cheers,
Max

dgalaktionov commented:
Thank you for the quick response.

How about moving all the parsing (and resolving) logic to a `Resolver`? It would be constructed with a string and have a `get` method that would return a vector of `Endpoint` objects. That way only the `Resolver` has to know about `std::async` and `std::future`, and we would also make `Endpoint` simple and stupid. In the future, this `Resolver` could be useful not only for a client but also for a server, if we want to allow the user to choose their listening interface by setting an IP address (I think Minecraft does that).

We could further separate the logic into `IPv6Resolver`, `IPv4Resolver` and `HostnameResolver`, which could all be invoked by a `CascadeResolver`, but that's a question of how much overengineered OO design you can tolerate in your codebase 😁

yamashi commented:
I like the idea of using a Resolver and then maybe instead of a "get" method, add operator[] and iterators ?

The IPv4, IPv6 etc seems a bit too much ^^ We could template it though in case someone really only wants IPv4/v6 or both.

dgalaktionov committed:
Fix build on linux

Turns out that for linux I also need to import netdb.h for getaddrinfo.

dgalaktionov committed:
Fix bad IPv4 test case for Linux

For some reason Linux (or at least my Linux) accepts 127.0.1 (and even 127.1) as valid and translates it to 127.0.0.1. Changed it to 127.0.0.0.1, which fails in both systems.

maximegmd commented on Code/tests/src/network.cpp:
```diff
@@ -1,331 +1,331 @@
-#include "catch.hpp"
```
Did you change the line endings by any chance ?

dgalaktionov replied:
Good catch. I had autocrlf = input in my linux .gitconfig
Won't happen again!

dgalaktionov committed:
Fix line endings

I've let my git replace CRLF endings to LF in my last 2 commits...

dgalaktionov committed:
Resolver class

dgalaktionov committed:
Some resolver tests

dgalaktionov committed:
Resolver acces and iterator methods

TIL that std::iterator was deprecated

dgalaktionov committed:
Remove Endpoint string constructor

Endpoint must not know anything about strings nor parsing now.

dgalaktionov committed:
Fix build on Linux (again)

Turns out I was too eager to remove #include <string> as it was needed for memcpy

dgalaktionov commented:
Should be good now. I've removed the string constructor and parsing functions from `Endpoint`, as this kind of initialization must now me done through `Resolver`. I've tried to follow the guidelines this time too.

maximegmd commented on Code/network/src/Resolver.cpp:
```diff
@@ -0,0 +1,148 @@
+
+std::vector<Endpoint> & Resolver::GetEndpoints() noexcept
+{
+    if (m_futureEndpoints.valid()) {
```
{ needs to be on a new line

maximegmd commented on Code/network/src/Resolver.cpp:
```diff
@@ -0,0 +1,148 @@
+void Resolver::Parse(const std::string & acAddress) noexcept
+{
+    uint16_t port = 0;
+    std::string strAddress(acAddress);
```
Might as well just pass the argument by value, this way we can get a move optimization

dgalaktionov replied:
Exactly my thought. I just copied it from your code and preferred not to touch it. Glad that you want it fixed 😄

maximegmd commented on Code/network/src/Resolver.cpp:
```diff
@@ -0,0 +1,148 @@
+
+Endpoint Resolver::GetEndpoint(const size_t & acIndex) noexcept
+{
+    return GetEndpoints().at(acIndex);
```
Don't use at(), it throws and I would love for the code to be exception free

dgalaktionov commented on Code/network/include/Resolver.h:
```diff
@@ -0,0 +1,53 @@
+        using reference = const Endpoint&;
+        using iterator_category = std::forward_iterator_tag;
+
+        Iterator() : m_iterator() {}
```
Should these braces go on new lines too?

maximegmd replied:
No it's fine for empty stuff like that

dgalaktionov committed:
Replace vector::at by [] operator

Also formatting changes

maximegmd commented on Code/network/src/Resolver.cpp:
```diff
@@ -64,52 +65,51 @@ std::vector<Endpoint> & Resolver::GetEndpoints() noexcept
 }
 
-void Resolver::Parse(const std::string & acAddress) noexcept
+void Resolver::Parse(std::string aAdress) noexcept
```
address takes 2 d ^^

dgalaktionov committed:
Fix typo

CAN I HAS ADRES
